salt/utils/subprocess.py

```diff
@@ -4,12 +4,53 @@
 '''
 from __future__ import absolute_import, print_function, unicode_literals
 
+import os
+import sys
 import shlex
+import logging
 import subprocess
 import threading
 import salt.exceptions
 import salt.utils.data
 from salt.ext import six
+from salt.ext.six.moves import range
+
+
+log = logging.getLogger(__name__)
+
+
+class FdPopen(subprocess.Popen):
+    '''
+    Python2 closing file descriptors in a blind way by a system's given range.
+    Therefore close_fds can significantly slow down the system in various circumstances.
+    '''
+
+    if sys.platform.startswith('linux') and six.PY2:
+        def _close_fds(self, but):
+            '''
+            Close only file descriptors that are needed.
+            The issue is not new: https://bugs.python.org/issue1663329
+
+            :param but:
+            :return:
+            '''
+            proc_path = '/proc/{}/fd'.format(os.getpid())
+            try:
+                maxfds = len(os.listdir(proc_path))
+            except OSError:
+                maxfds = subprocess.MAXFD
+
+            log.debug('Closing %s file descriptors', maxfds)
+            if hasattr(os, 'closerange'):
+                os.closerange(3, but)
+                os.closerange(but + 1, maxfds)
+            else:
+                for i in range(3, maxfds):
```

Comment:
`salt.ext.six.moves.range` should be used here instead of `range`

Reply:
Yeah, this makes range use xrange on py2 so that it's a generator. You can import it like so:

```python
from salt.ext.six.moves import range  # pylint: disable=import-error,redefined-builtin
```

Reply:
@DmitryKuzmenko but this is what I am doing, if you had checked the imports part of this file:
```python
from salt.ext.six.moves import range
```

Reply:
@terminalmage I have no pylint disable there though. Is it needed?

Reply:
Ah. Sorry, I missed this fact. Just because usually we do `from salt.ext import six` and use `six.somehting` to not mess it with builtins.